### PR TITLE
[tests-only] Adjust pact interaction of sharingTest

### PR DESCRIPTION
**Pull request by saw-jan**
part of https://github.com/owncloud/owncloud-sdk/issues/709 and https://github.com/owncloud/owncloud-sdk/issues/644

adjustment for test suites:
- sharingTest

### CODE REVIEW

File: tests/sharingTest.js [hunk 1]
```diff
@@ -35,175 +37,480 @@ describe('Main: Currently testing file/folder sharing,', function () {
     '文件.txt': 19
   }
   const testFolderShareID = 9
+  const testFolderShareToken = '6Bvt5fDajGxiV3P'
 
   // CONSTANTS FROM lib/public/constants.php
   const OCS_PERMISSION_READ = 1
   const OCS_PERMISSION_UPDATE = 2
-  const OCS_PERMISSION_CREATE = 4
+  // const OCS_PERMISSION_CREATE = 4
```
Comment (jasson99) [inline tests/sharingTest.js:45]: Can this be removed then?

File: tests/sharingTest.js [hunk 2]
```diff
@@ -35,175 +37,480 @@ describe('Main: Currently testing file/folder sharing,', function () {
...
+      await givenUserExists(provider, testUser, testUserPassword)
+      return givenResourceShared(username, password, provider, 0, resource, testUser)
+    } else if (shareType === 1) {
+      await givenGroupExists(provider, testGroup)
+      return givenResourceShared(username, password, provider, 1, resource, testGroup)
+    }
+  }
+  const initFileFolderUserGroupShare = async (provider, shareType, resource, resourceType) => {
+    await createFileOrFolder(adminUser, adminPassword, provider, resource, resourceType)
+    return createShare(adminUser, adminPassword, provider, shareType, resource)
```
Comment (jasson99) [inline tests/sharingTest.js:94]: These should be performed by test user rather than admin user

**Review by jasson99 — CHANGES_REQUESTED** [review]
The requests should be performed by test user rather than admin user

File: tests/sharingTest.js [hunk 3]
```diff
@@ -1,224 +1,538 @@
...
   const testFiles = config.testFiles
 
+  const sharer = config.testUser
+  const sharerPassword = config.testUserPassword
+  const sharee = config.testUser2
+  const shareePassword = config.testUser2Password
+
+  const validAuthHeaders = {
+    authorization: getAuthHeaders(sharer, sharerPassword)
+  }
```
Comment (jasson99) [inline tests/sharingTest.js:39]: may be we can shift this in pactHelper and can use in other files as well, as most of them require validUserAuthHeaders now.

**Review by jasson99 — COMMENTED** [review]
minor suggestions only

File: tests/sharingTest.js (hunk 3 above)
Comment (saw-jan) [reply tests/sharingTest.js:39]: we can implement something like this
```js
const getValidUserAuthHeaders = (username, password) => ({
  authorization: getAuthHeaders(username, password)
})
```
and use it like
```js
// on every interaction
...
headers: getValidUserAuthHeaders(username, password)
...

OR

validUserAuthHeaders = getValidUserAuthHeaders(username, password)
...
headers: validUserAuthHeaders
...
```

It is also possible to introduce some confusion between `getAuthHeaders` and `getValidUserAuthHeaders`, and workarounds

File: tests/providerTest.js [hunk 4]
```diff
@@ -144,6 +149,32 @@ describe('provider testing', () => {
...
       'provider base url is returned': () => {
         return { providerBaseURL: process.env.PROVIDER_BASE_URL }
+      },
+      'resource is shared': (setup, parameters) => {
+        if (setup) {
+          const { username, password, resource, shareType, shareWith } = parameters
+          const response = shareResource(username, password, resource, shareType, shareWith)
+
+          const { status } = getOCSMeta(response)
+          /* eslint-disable-next-line */
```
Comment (dpakach) [inline tests/providerTest.js:159]: please disable only one rule that's causing error

Comment (saw-jan) [reply tests/providerTest.js:159]: this is to disable the following one line,
how can I disable a specific rule?

Comment (individual-it) [reply tests/providerTest.js]: e.g `// eslint-disable-next-line no-useless-escape`

File: tests/helpers/sharingHelper.js [hunk 5]
```diff
@@ -147,9 +147,38 @@ const createFileInLastPublicShare = function (token, fileName) {
...
+    }
+    if (obj.password) {
+      toUpdate = `password=${obj.password}`
+    }
+    if (obj.expireDate) {
+      toUpdate = `expireDate=${obj.expireDate}`
+    }
+    if (obj.publicUpload !== undefined && obj.publicUpload !== null && obj.publicUpload !== '') {
+      toUpdate = `publicUpload=${obj.publicUpload.toString().toLowerCase()}`
+    }
```
Comment (dpakach) [inline tests/helpers/sharingHelper.js:173]: how about just loop through the object and append, to make it more cleaner

File: tests/sharingTest.js [hunk 6]
```diff
@@ -1,33 +1,41 @@
...
+  const testFilesId = config.testFilesId
+
+  const sharer = config.testUser
+  const sharerPassword = config.testUserPassword
+  const sharee = config.testUser2
+  const shareePassword = config.testUser2Password
+
+  const validAuthHeaders = {
+    authorization: getAuthHeaders(sharer, sharerPassword)
+  }
```
Comment (jasson99) [inline tests/sharingTest.js:38]: IMO, this can be used in other files as well, so we can keep this in pact helper directly, with getAuthHeaders, validAdminAuthHeaders. Using it this way can be uniform when we can use in other files as well.

**Review by jasson99 — COMMENTED** [review]
Just minor suggestion. Other than that, Looks good to me

File: tests/sharingTest.js (hunk 6 above)
Comment (saw-jan) [reply tests/sharingTest.js:38]: we can make it at once for all tests in another PR

**Review by swoichha — COMMENTED** [review]
lgtm

File: tests/helpers/sharingHelper.js [hunk 7]
```diff
@@ -147,9 +136,31 @@ const createFileInLastPublicShare = function (token, fileName) {
...
 }
 
+/**
+ * convert JS object to form url encoded data.
+ *
+ * @param {object} obj
+ * @returns {string}
+ */
+const toFormUrlEncoded = function (obj) {
+  let toUpdate
```
Comment (individual-it) [inline tests/helpers/sharingHelper.js:146]: that variable name somehow does not have anything to do with the function name. maybe call it something like `formData`

**Review by Talank — COMMENTED** [review]
LGTM

File: tests/sharingTest.js [hunk 8]
```diff
@@ -36,190 +44,494 @@ describe('Main: Currently testing file/folder sharing,', function () {
...
+
+  describe('Currently testing folder sharing,', function () {
+    describe('sharedFolderByLink,', function () {
+      describe('checking the shared folder,', function () {
+        it('checking method : isShared with shared folder', async function () {
+          const provider = createProvider()
+          await getCapabilitiesInteraction(provider, sharer, sharerPassword)
+          await getCurrentUserInformationInteraction(provider, sharer, sharerPassword)
+          await getIsSharedInteraction(provider, `check whether '${testFolder}' is shared or not (public link share)`, 3, testFolder, 'folder')
```
Comment (individual-it) [inline tests/sharingTest.js:272]: pretty long line

File: tests/sharingTest.js [hunk 9]
```diff
@@ -36,190 +44,494 @@ describe('Main: Currently testing file/folder sharing,', function () {
...
+  const givenFolderExists = (provider, username, password, resource) => {
+    return provider
+      .given('folder exists', { username, password, folderName: resource })
+  }
+  const givenResourceShared = async (provider, username, password, shareType, resource, shareWith) => {
+    return provider
+      .given('resource is shared', { username, password, resource, shareType, shareWith })
+  }
+  let shareId, shareToken
+  const createFileOrFolder = (provider, username, password, resource, resourceType) => {
```
Comment (individual-it) [inline tests/sharingTest.js:75]: lets be consistent with the naming
```suggestion
  const givenfFileFolderExists = (provider, username, password, resource, resourceType) => {
```

File: tests/sharingTest.js [hunk 10]
```diff
@@ -36,190 +44,494 @@ describe('Main: Currently testing file/folder sharing,', function () {
...
+      shareToken = testFolderShareToken
+      return givenFolderExists(provider, username, password, resource)
+    } else {
+      const iteration = testFiles.indexOf(resource)
+      shareId = config.testFilesId[iteration]
+      shareToken = config.testFilesToken[iteration]
+      return givenFileExists(provider, username, password, resource)
+    }
+  }
+  const createShare = async (provider, username, password, shareType, resource) => {
```
Comment (individual-it) [inline tests/sharingTest.js:87]: make a consistent name with `given`

File: tests/sharingTest.js [hunk 11]
```diff
@@ -36,190 +44,494 @@ describe('Main: Currently testing file/folder sharing,', function () {
...
+        it('adding password', async function () {
+          const toUpdate = { password: '1234' }
+          const additionalBodyElement = {
+            share_with: testGroup,
+            share_with_displayname: testGroup
+          }
+          const provider = createProvider()
+          await getCapabilitiesInteraction(provider, sharer, sharerPassword)
+          await getCurrentUserInformationInteraction(provider, sharer, sharerPassword)
+          await updateShareInteraction(provider, '(group share: set share password)', 1, testFolder, 'folder', toUpdate, additionalBodyElement)
```
Comment (individual-it) [inline tests/sharingTest.js:487]: too long line

File: tests/sharingTest.js [hunk 12]
```diff
@@ -710,19 +956,21 @@ describe('Main: Currently testing file/folder sharing,', function () {
...
-        await shareGETExistingNonSharedFile(provider)
+        await shareGETExistingNonSharedFile(provider, '', fileName)
 
         return provider.executeTest(async () => {
-          const oc = createOwncloud()
+          const oc = createOwncloud(sharer, sharerPassword)
           await oc.login()
-          const suffix = '123'
-          return oc.files.putFileContents('newFileCreated' + suffix).then(status => {
+          return oc.files.putFileContents(fileName).then(status => {
```
Comment (individual-it) [inline tests/sharingTest.js:971]: we don't need to write the file here, because this is not about putFileContents but about testing the consumer side of `isShared`
but let's leave it as it is here for now

**Review by individual-it — CHANGES_REQUESTED** [review]

File: tests/helpers/sharingHelper.js [hunk 13]
```diff
@@ -147,9 +136,31 @@ const createFileInLastPublicShare = function (token, fileName) {
   })
 }
 
+/**
+ * convert JS object to form url encoded data.
+ *
+ * @param {object} obj
+ * @returns {string}
+ */
+const toFormUrlEncoded = function (obj) {
```
[the remark this answers is not in the document]
Comment (saw-jan) [reply tests/helpers/sharingHelper.js:145]: it is mainly used to convert a share's update parameter which is in js object into `key=value` pair
As only one update parameter works per request, I made a break

**Review by individual-it — APPROVED** [review]

File: tests/sharingTest.js [hunk 14]
```diff
@@ -408,83 +736,71 @@ describe('Main: Currently testing file/folder sharing,', function () {
...
-          for (file in sharedFiles) {
-            await oc.shares.getShares(file).then(shares => {
-              expect(shares.constructor).toBe(Array)
-              let flag = 0
-              for (let i = 0; i < shares.length; i++) {
-                const share = shares[i]
-                if (allIDs.indexOf(share.getId()) > -1) {
-                  flag = 1
-                }
-              }
```
Comment (dpakach) [inline tests/sharingTest.js:458]: why did you remove this part of the test

File: tests/sharingTest.js [hunk 15]
```diff
@@ -332,71 +685,46 @@ describe('Main: Currently testing file/folder sharing,', function () {
...
                 expect(typeof (share)).toBe('object')
                 expect(share.getId()).toBeGreaterThan(-1)
               }).catch(error => {
                 expect(error).toBe(null)
               })
-            }
-          })
-        })
-
-        it('getShares for shared file', async function () {
```
Comment (dpakach) [inline tests/sharingTest.js:355]: why is this test removed?

File: tests/sharingTest.js [hunk 16]
```diff
@@ -232,95 +600,80 @@ describe('Main: Currently testing file/folder sharing,', function () {
           })
         })
 
-        it('checking method : getShares for shared file', async function () {
```
Comment (dpakach) [inline tests/sharingTest.js:235]: is this test removed?

Comment (individual-it) [reply tests/sharingTest.js:235]: isn't is that here https://github.com/owncloud/owncloud-sdk/pull/728/files#diff-3e96a43471ac474278d7585f7a1ab0a201ce7c5a935710b88c2a9304710d8a76R763 ?

Comment (saw-jan) [reply tests/sharingTest.js:235]: that test is in this line https://github.com/owncloud/owncloud-sdk/blob/a4355e367127b9412421e5d8a193c6499c0757f0/tests/sharingTest.js#L763

File: tests/sharingTest.js (hunk 14 above)
Comment (saw-jan) [reply tests/sharingTest.js:458]: I directly counted share count and asserted it, because at given state three shares are created

**Review by dpakach — APPROVED** [review]
LGTM :+1:
